src/P3Scheme.jl

```diff
@@ -667,4 +767,464 @@ function D_m(p3::PSP3, q::FT, N::FT, ρ_r::FT, F_r::FT) where {FT}
+
+ - p3 - a struct containing P3 Scheme parameters
+ - Chen2022 - a struct with terminal velocity parameters as in Chen(2022)
+ - ρ_a - density of air  
+ - D - maximum particle dimension
+ - F_r - rime mass fraction (q_rim/q_i)
+ - th - P3Scheme nonlinear solve output tuple (D_cr, D_gr, ρ_g, ρ_d)
+
+ Returns the terminal velocity of ice at given particle dimension using P3 
+ distributions and Chen 2022 parametrizations
+"""
+function velocity_chen(
+    D::FT,
+    p3::PSP3,
+    Chen2022::CMP.Chen2022VelTypeSnowIce,
+    ρ_a::FT,
+    F_r::FT,
+    th = (; D_cr = FT(0), D_gr = FT(0), ρ_g = FT(0), ρ_d = FT(0)),
+) where {FT}
+    if D <= FT(0.000625)
+        (ai, bi, ci) = CO.Chen2022_vel_coeffs_small(Chen2022, ρ_a)
+    else
+        (ai, bi, ci) = CO.Chen2022_vel_coeffs_large(Chen2022, ρ_a)
+    end
+
+    κ = FT(-1 / 6)
+
+    v = 0
+    for i in 1:2
+        v += ai[i] * D^(bi[i]) * exp(-ci[i] * D)
+    end
+
+    return ϕ_ice(D, p3, F_r, th)^κ * v
+end
+
+"""
+    velocity_chen(p3, Chen2022, ρ_a, D)
+
+ - Chen2022 - a struct with terminal velocity parameters as in Chen(2022)
+ - ρ_a - density of air  
+ - D - maximum particle dimension
+ 
+ Returns the terminal velocity of rain given Chen 2022 velocity parametrizations
+"""
+function velocity_chen(
+    D::FT,
+    Chen2022::CMP.Chen2022VelTypeRain,
+    ρ_a::FT,
+) where {FT}
+    (ai, bi, ci) = CO.Chen2022_vel_coeffs_small(Chen2022, ρ_a)
+
+    v = 0
+    for i in 1:3
+        v += ai[i] * D^(bi[i]) * exp(-ci[i] * D)
+    end
+
+    return v
+end
+
+function get_ice_bound(p3, λ::FT, tolerance::FT) where {FT}
+    ice_problem(x) =
+        tolerance - Γ(1 + DSD_μ(p3, λ), FT(exp(x)) * λ) / Γ(1 + DSD_μ(p3, λ))
+    guess = log(19 / 6 * (DSD_μ(p3, λ) - 1) + 39) - log(λ)
+    log_ice_x =
+        RS.find_zero(
+            ice_problem,
+            RS.SecantMethod(guess - 1, guess),
+            RS.CompactSolution(),
+            RS.RelativeSolutionTolerance(eps(FT)),
+            5,
+        ).root
+
+    return exp(log_ice_x)
+end
+
+"""
+    integration_bounds(p3, tolerance, λ, N_0, Nᵢ, qᵣ, Nᵣ)
+
+ - p3 - a struct containing P3 Scheme parameters 
+ - tolerance - tolerance to which distributions need to be evaluated to
+ - λ - shape parameter of ice distribution
+ - N_0 - intercept size distribution of ice 
+ - Nᵢ - number mixing ratio of ice
+ - q - mass mizing ratio of colliding species
+ - N - number mixing ratio of colliding species
+
+ Returns the bounds over which to integrate rain, cloud, and ice distributions to ensure 
+ coverage or more than (1 - tolerance) of each distribution
+"""
+function integration_bounds(
+    type::String,
+    p3::PSP3,
+    tolerance::FT,
+    λ::FT,
+    N_0::FT,
+    Nᵢ::FT,
+    q::FT,
+    N::FT,
+) where {FT}
+    if type == "rain"
+        rain_λ, = get_rain_parameters(q, N)
+        colliding_x = -1 / rain_λ * log(tolerance)
+    elseif type == "cloud"
+        cloud_λ, = get_cloud_parameters(q, N)
+        cloud_problem(x) =
```

Comment:
The same for th bounds for the `cloud_probelm`. Could you explain how you get those bounds? Could the logic for them be combined between the cloud rain and ice?